### PR TITLE
Allow piecewise signatures in records definition

### PR DESCRIPTION
**Pull request by litchipi**
Implementation of #496.
Merged `TypeAnnot` into `MetaAnnot` (renamed now as `Annot`) inside `grammar.lalrpop`.
This allows that kind of Nickel code inside records:
```
{
   foo : Num,
   bar = 3,
   foo = 5
}
```
It could also let these partial type-annotations to be used when destructuring a record like so: `let {foo : Num, bar} = x`
However the semantic of this specific case needs to be precised.

> Wether or not to allow undefined type-annotated fields are yet to be decided, and are allowed for now as this is a pre-relase.
> Wether or not to allow code in the middle of a field's type annotation and definition is yet to be decided also.
> Both these points should be debated in a separated issue before modifying this feature.

### CODE REVIEW

**Review by yannham — COMMENTED** [review]
A couple tests for other edge cases would be nice as well:

Test that this fails (in tyepcheck_fail.rs e.g.)
```
{
  foo : Num,
  foo = "bar",
}
```

Test that this passes
```
{
  foo : Num,
  foo = 1,
  bar : Num = foo,
}

// Test that this parses fine:
{foo : Num}

// Test that this evaluates fine
let {foo : Num} = {foo = 1} in foo
```

 (suggestions)

**Comment by litchipi** [comment]
```
{
  foo : Num,
  foo = 1,
  bar : Num = foo,
}
```
This new test doesn't pass as it tries to unify `Dyn` with `Num`.
In the internals, the record is considered as a `RecRecord` and that makes the evaluation use `Dyn` everywhere.
I suppose it's not a normal behavior, it should be considered as a special form of static record, what do you think about it ?

File: src/grammar.lalrpop [hunk 1]
```diff
@@ -278,15 +252,15 @@ Destruct: Destruct = {
 };
 
 Match: Match = {
-    <left:Ident> <anns: MetaAnnot?> <default: DefaultAnnot?> "=" <right: Pattern> => {
+    <left:Ident> <anns: Annot?> <default: DefaultAnnot?> "=" <right: Pattern> => {
```
[the remark this answers is not in the document]
Comment (litchipi) [reply src/grammar.lalrpop:255]: Makes sense to me

**Comment by yannham** [comment]
It shouldn't be related to the type of record. The type of each field is dictated by the content: here, we want the content to be a metavalue with a `type` field set to `Num` in both places. If that's not the case there's something fishy going on with respect to `flatten` or the way `build_record` operate.

**Comment by yannham** [comment]
What is possible is that until now, `build_record` just merged everything dynamically without care. If that's the case, we probably rather want to gather all the values first and statically flatten the outer metavalues.

**Comment by litchipi** [comment]
It seams that this code:
```
{
  foo : Num,
  foo = 1,
  bar : Num = foo,
}
```
makes Nickel use an `Op2::Merge` operation over `foo` to create a single term, no problem from here.
However in `get_bop_type()`, this operation returns `(Dyn, Dyn, Dyn)` types which enters in conflict with the statically typed `bar : Num`, and causes the bug.

Is the way to go is simply to avoid using `Op2::Merge` ? Perform the merge much sooner, in the parser ?

**Comment by yannham** [comment]
> Is the way to go is simply to avoid using Op2::Merge ? Perform the merge much sooner, in the parser ?

Yes, precisely. At least for the *outer metavalues*. I guess it's fine to rely on merge for the inner values. So, for example, if you have a vector of definitions for field `foo`, you would morally split them as `(Option<MetaData>, RichTerm)`, flatten the metadata, merge the terms, and reconstruct a `MetaValue` with the merged terms as inner term, or directly the merged terms if no definition had metavalues (it's just draft steps, not saying you should necessarily do exactly like that).

**Review by yannham — CHANGES_REQUESTED** [review]

**Review by yannham — CHANGES_REQUESTED** [review]

File: src/parser/utils.rs [hunk 2]
```diff
@@ -219,6 +220,55 @@ where
...
+        if let Some(t2) = term2.1 {
+            mk_term::op2(BinaryOp::Merge(), t1, t2)
+        } else {
+            t1
+        }
+    } else {
+        if let Some(t2) = term2.1 {
+            t2
+        } else {
+            RichTerm::from(Term::Null)
```
[the remark this answers is not in the document]
Comment (litchipi) [reply src/parser/utils.rs:247]: I made the function return an `Option<RichTerm>` which I unwrap in the `insert_static_field` as it's impossible to get a field with `value = None` AND `metavalue = None`.

I didn't unwrap it inside the function as it could be reused for other usages later from which this particular case may not be impossible (I guess ?)

**Comment by litchipi** [comment]
I got a problem creating the `test/merge_fail.rs` file. I created a macro:
```
macro_rules! assert_merge_fails {
    ($term:expr) => {
        assert_matches!(
            eval($term),
            Err(Error::EvalError(EvalError::MergeIncompatibleArgs(..)))
        )
    };
}
```
(inspired from `src/eval/test.rs` and `test/typecheck_fail.rs`)
And this happends:
```
// Works
assert_merge_fails!("({a=1} & {a=2}).a");  
// Doesn't work
assert_merge_fails!("{ foo = (fun x => x) (1 | default), foo = (fun x => x) (1 | default) } & {foo | default = 2 }");   
```
Even tho when I write the exact same test in a `a.ncl` file and `cargo run` it, it fails like it should.
Is there anything different from using the commandline and using the `eval` function (from the `utilities` crate) ?

**Comment by yannham** [comment]
>Even tho when I write the exact same test in a a.ncl file and cargo run it, it fails like it should.
Is there anything different from using the commandline and using the eval function (from the utilities crate) ?

Yes, we recently made the default `nickel` command and the `repl` to deeply evaluate the argument. This it triggers the contract check. On the other hand, you are just using the standard `eval`, that doesn't evaluate further than the first record. You need to either `eval_full`, or you can also access the field `foo`:

```
({ foo = (fun x => x) (1 | default), foo = (fun x => x) (1 | default) } & {foo | default = 2 }).foo
```

to trigger the error.

**Comment by litchipi** [comment]
Yep indeed, should have found that ^^'
Pushed the last version, cleaned and ready to be merged if it's OK for you

**Review by yannham — APPROVED** [review]
LGTM. But you'll have to rebase because the `Rc` PR has been merged.

**Comment by yannham** [comment]
@litchipi is there anything to do before merging this?

**Comment by litchipi** [comment]
Nope, not anymore on my side I think it can be merged
